openedx/features/course_experience/tests/views/test_course_home.py

```diff
@@ -251,6 +252,35 @@ def test_home_page(self, user_type, expected_message):
         if expected_message:
             self.assertContains(response, expected_message)
 
+    @ddt.data(
+        [False, 'private', 'to see course content'],
+        [False, 'preview', 'to see course content'],
+        [False, 'public', 'to see course content'],
+        [True, 'private', 'to see course content'],
+        [True, 'preview', TEST_CHAPTER_NAME],
+        [True, 'public', TEST_CHAPTER_NAME],
+    )
+    @ddt.unpack
+    def test_home_anonymous_user(self, waffle_override, course_visibility, expected_message):
```

Comment:
Can this not be merged with `test_home_page()` so that status of all sub-fragments can be tested? For example, the 3 ddt variations with logged in users can have the flag `False` and these 6 variations can have `CourseUserType.ANONYMOUS`.

Reply:
Also instead of patching see the `update_item(self.course, self.user.id)` pattern in `courseware/tests/test_views.py` for updating the course module fields.

Reply:
I tried merging at first but the test was getting very complex, so I decided to keep those separate.

I moved to patching as updating the course module was causing troubles on other tests, even when rolledback the changes.

Reply:
Since the existing test checks the visibility of the different fragments, it makes sense to have one which checks all the permutations. The new test is not checking the visibility of the other fragments. The main changes I see are:

1. `test_home_page(self, enable_anonymous_courseware_access, course_visibility, user_type, expected_message)`
2. The `patch` and `override_waffle_flag` need to go around the `self.client.get()` as in the new test.
3. Updating one assert:
```
self.assertContains(response, TEST_CHAPTER_NAME, count=1 if (is_enrolled or is_unenrolled_staff or (enable_anonymous_courseware_access and course_visiblity in ['preview', 'public'])) else 0)
```

What else needs to change?

Reply:
Makes sense, I also had to change the expected number of welcome messages.